### PR TITLE
Enable feature flagging on Postgres schemas

### PR DESCRIPTION
**Pull request by connorgorman**
## Description

Allow for feature flagging schemas so that they can be in development for longer than a release without impacting Postgres

## Checklist
- [ ] Investigated and inspected CI test results
- [ ] Unit test and regression tests added
- [ ] Evaluated and added CHANGELOG entry if required
- [ ] Determined and documented upgrade steps
- [ ] Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))

If any of these don't apply, please comment below.

## Testing Performed

Ran this on the blob store PR

### CODE REVIEW

File: tools/generate-helpers/pg-table-bindings/store_test.go.tpl [hunk 1]
```diff
@@ -43,9 +44,16 @@ func (s *{{$namePrefix}}StoreSuite) SetupSuite() {
...
+	s.T().Setenv(features.{{.FeatureFlag}}.EnvVar(), "true")
+	if !features.{{.FeatureFlag}}.Enabled() {
+		s.T().Skip("Skip postgres store tests because feature flag is off")
+		s.T().SkipNow()
+	}
+	{{- end }}
 
 	s.testDB = pgtest.ForT(s.T())
-	s.store = New(s.testDB.DB)
+	s.store = CreateTableAndNewStore(context.Background(), s.testDB.DB, s.testDB.GetGormDB(s.T()))
```
Comment (connorgorman) [inline tools/generate-helpers/pg-table-bindings/store_test.go.tpl:56]: This is required because of timing

Comment (connorgorman) [reply tools/generate-helpers/pg-table-bindings/store_test.go.tpl:56]: Wait, maybe this is not needed now

File: pkg/postgres/schema/all.go [hunk 2]
```diff
@@ -21,17 +21,22 @@ var (
...
-	Schema     *walker.Schema
-	CreateStmt *postgres.CreateStmts
+	Schema          *walker.Schema
+	CreateStmt      *postgres.CreateStmts
+	FeatureFlagFunc func() bool
 }
 
 // RegisterTable maps a table to an object type for the purposes of metrics gathering
-func RegisterTable(schema *walker.Schema, stmt *postgres.CreateStmts) {
+func RegisterTable(schema *walker.Schema, stmt *postgres.CreateStmts, featureFlagFuncs ...func() bool) {
```
Comment (rhmdnd) [inline pkg/postgres/schema/all.go:30]: Interesting. So we could use this to register new schemas within the code, but would there be a corresponding migration script?

I was looking through the migration scripts/documentation earlier and noticed they were applied sequentially, but this doesn't appear to be related to those migrations.

Comment (rhmdnd) [reply pkg/postgres/schema/all.go:30]: Would we keep this opt-in registration specific to new tables? For example, would this evolve to support optional schemas that aren't table additions?

File: tools/generate-helpers/pg-table-bindings/store_test.go.tpl (hunk 1 above)
Comment (c-du) [reply tools/generate-helpers/pg-table-bindings/store_test.go.tpl]: Weird, what is the timing issue? I remember all the schema are applied for store tests.

File: pkg/postgres/schema/all.go [hunk 3]
```diff
@@ -21,17 +21,22 @@ var (
 )
 
 type registeredTable struct {
-	Schema     *walker.Schema
-	CreateStmt *postgres.CreateStmts
+	Schema          *walker.Schema
+	CreateStmt      *postgres.CreateStmts
+	FeatureFlagFunc func() bool
```
Comment (c-du) [inline pkg/postgres/schema/all.go:26]: How about FeatureEnabledFunc or FeatureEnabledLamda :)

File: pkg/postgres/schema/all.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (connorgorman) [reply pkg/postgres/schema/all.go:30]: I wanted to avoid changing all of the generated files

**Review by c-du — APPROVED** [review]

File: pkg/postgres/schema/all.go (hunk 2 above)
Comment (connorgorman) [reply pkg/postgres/schema/all.go:30]: So the schemas are created and migrated at the end of the migrator. In this case, this schema would not be created

Comment (dashrews78) [reply pkg/postgres/schema/all.go:30]: Maybe the migration conversation could be taken off line, but how would this work for a migration?  The fact they run in sequence would make it difficult to put a migration behind a gate OR is the idea we are only going to write the migration at the end when we are ready to remove the gate?

Comment (connorgorman) [reply pkg/postgres/schema/all.go:30]: This is a good question, but we'll need to figure out if we can effectively flag migrations

**Review by md2119 — COMMENTED** [review]
Currently, without feature flagging, we have to be thorough about making changes to the schema.  Especially if the development spans across multiple releases. For instance, if we add a column and want to remove it later or vice-versa, it results in multiple migrations even though the feature was not production-ready.

I see the benefit of the feature flagging the registration of the whole schema. However, it seems to be that if users want to use schema A when a feature is disabled and schema B if the feature flag is enabled, two schemas must exist somewhere. One, obviously, must reside where active schema resides, but the other somewhere else. That means we have to be careful and make sure we refer to the correct schema when we want to use them.
